spec/models/odk/response_saver_spec.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+require "rails_helper"
+
+describe ODK::ResponseSaver do
+  include_context "odk submissions"
+
+  let(:mission) { create(:mission) }
+  let(:user) { create(:user, role_name: "enumerator", mission: mission) }
+  let(:fixture_name) { "single_question" }
+  let(:xml) { prepare_odk_response_fixture(fixture_name, form, values: [1], formver: formver) }
+  let(:file) { Tempfile.new.tap { |f| f.write(xml) && f.rewind } }
+  let(:upload) { Rack::Test::UploadedFile.new(file, "text/xml") }
+  let(:request_params) { {xml_submission_file: upload, format: "xml"} }
+  let(:nemo_response) { Response.first }
+  let(:save_fixtures) { true }
+
+  context "race conditions" do
+    let(:xml_values) { %w[A B C D] }
+    let!(:form1) { create(:form, :live, mission: mission, question_types: question_types) }
+    let!(:formver) { create(:form_version, code: "abc", number: "202211", form: form1) }
+    let(:r1_path) { "tmp/odk/responses/simple_response/simple_response.xml" }
+    let(:r1) do
+      build(
+        :response,
+        :with_odk_attachment,
+        xml_path: r1_path,
+        form: form1,
+        answer_values: xml_values,
+        user_id: user.id
+      )
+    end
+    let!(:question_types) { %w[text text text text] }
+
+    it "should return a database serialization error", database_cleaner: :truncate do
+      # make odk
+      prepare_odk_response_fixture("simple_response", form1, values: xml_values, formver: "202211")
+      r1_path = Rails.root.join("tmp/odk/responses/simple_response/simple_response.xml")
+      upload = Rack::Test::UploadedFile.new(r1_path, "text/xml")
+      checksum = ODK::ResponseParser.compute_checksum_in_chunks(upload)
+      e = nil
+
+      thread1 = Thread.new do
+        e = ODK::ResponseSaver.save_with_retries!(
+          response: r1,
+          submission_file: upload,
+          user_id: user.id,
+          test: true
+        )
+      end
+
+      thread2 = Thread.new do
+        # wait until the first thread is sleepin
+        sleep(2)
+        insert_response_via_second_db_connection(checksum)
+      end
+
+      thread1.join
+      thread2.join
+
+      expect(ActiveStorage::Blob.all.count).to eq(1)
+      expect(e.class).to eq(ActiveRecord::SerializationFailure)
+    end
+  end
+
+  def insert_response_via_second_db_connection(checksum)
```

Comment:
There isn't a way of simulating a new rails workers, but you can simulate a new database thread which accomplishes the same thing.

Reply:
This method also just inserts the blob, and doesn't insert the response because we are testing duplicates on the checksum which is on the blob.